File: android/GT7Collector/app/src/main/java/com/trc/gt7collector/TimingApi.kt
package com.trc.gt7collector

import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class TimingApi {
    fun fetchEntries(serverUrl: String, raceCode: String): CollectorRaceResponse {
        val code = encode(raceCode.trim().uppercase())
        val json = request("GET", "$serverUrl/api/collector/races/$code/entries")
        val race = json.getJSONObject("race")
        val entriesJson = json.getJSONArray("entries")
        val entries = (0 until entriesJson.length()).map { index ->
            val entry = entriesJson.getJSONObject(index)
            val driversJson = entry.getJSONArray("drivers")
            EntryInfo(
                entryId = entry.getString("entry_id"),
                carNumber = entry.getInt("car_number"),
                teamName = entry.getString("team_name"),
                carModel = entry.getString("car_model"),
                carClass = entry.getString("class"),
                drivers = (0 until driversJson.length()).map { driverIndex ->
                    val driver = driversJson.getJSONObject(driverIndex)
                    DriverInfo(
                        driverId = driver.getString("driver_id"),
                        displayName = driver.getString("display_name")
                    )
                }
            )
        }
        return CollectorRaceResponse(
            raceId = race.getString("race_id"),
            raceName = race.getString("name"),
            entries = entries
        )
    }

    fun register(
        serverUrl: String,
        raceCode: String,
        entryId: String,
        teamCode: String,
        driverId: String
    ): String {
        val body = JSONObject().apply {
            put("race_code", raceCode.trim().uppercase())
            put("entry_id", entryId)
            put("team_code", teamCode)
            put("driver_id", driverId)
            put("collector_version", "android-0.1.0")
        }
        val response = request("POST", "$serverUrl/api/collector/register", body)
        if (!response.optBoolean("send_allowed", false)) {
            throw TimingApiException("Race Control does not allow this collector to send data.")
        }
        return response.getString("collector_token")
    }

    fun sendTelemetry(serverUrl: String, token: String, payload: TelemetryPayload) {
        request(
            method = "POST",
            urlText = "$serverUrl/api/collector/telemetry",
            body = payload.toJson(),
            token = token
        )
    }

    private fun request(
        method: String,
        urlText: String,
        body: JSONObject? = null,
        token: String? = null
    ): JSONObject {
        val url = URL(urlText)
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 8000
            readTimeout = 12000
            setRequestProperty("Accept", "application/json")
            if (token != null) {
                setRequestProperty("Authorization", "Bearer $token")
            }
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }

        if (body != null) {
            OutputStreamWriter(connection.outputStream, Charsets.UTF_8).use {
                it.write(body.toString())
            }
        }

        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val text = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        if (status !in 200..299) {
            val detail = runCatching { JSONObject(text).optString("detail") }.getOrNull()
            throw TimingApiException(detail?.takeIf { it.isNotBlank() } ?: "Timing server rejected the request ($status).")
        }
        return if (text.isBlank()) JSONObject() else JSONObject(text)
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, Charsets.UTF_8.name())
}

class TimingApiException(message: String) : RuntimeException(message)

File: ios/GT7Collector/GT7Collector/TimingAPI.swift
import Foundation

enum TimingAPIError: LocalizedError {
    case invalidServerURL
    case requestFailed(String)
    case serverRejected(String)

    var errorDescription: String? {
        switch self {
        case .invalidServerURL:
            return "Server URL is invalid."
        case .requestFailed(let message):
            return message
        case .serverRejected(let message):
            return message
        }
    }
}

final class TimingAPI {
    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    init(session: URLSession = .shared) {
        self.session = session
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
    }

    func fetchEntries(serverURL: String, raceCode: String) async throws -> CollectorRaceResponse {
        let url = try endpoint(serverURL, "/api/collector/races/\(raceCode.uppercased())/entries")
        let (data, response) = try await session.data(from: url)
        try validate(response, data: data)
        return try decoder.decode(CollectorRaceResponse.self, from: data)
    }

    func register(
        serverURL: String,
        raceCode: String,
        entryId: String,
        teamCode: String,
        driverId: String
    ) async throws -> CollectorRegisterResponse {
        let request = CollectorRegisterRequest(
            raceCode: raceCode.uppercased(),
            entryId: entryId,
            teamCode: teamCode,
            driverId: driverId,
            collectorVersion: "ios-0.1.0"
        )
        let url = try endpoint(serverURL, "/api/collector/register")
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try encoder.encode(request)
        let (data, response) = try await session.data(for: urlRequest)
        try validate(response, data: data)
        return try decoder.decode(CollectorRegisterResponse.self, from: data)
    }

    func sendTelemetry(serverURL: String, token: String, payload: TelemetryPayload) async throws {
        let url = try endpoint(serverURL, "/api/collector/telemetry")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try encoder.encode(payload)
        let (data, response) = try await session.data(for: request)
        try validate(response, data: data)
    }

    private func endpoint(_ serverURL: String, _ path: String) throws -> URL {
        guard let base = URL(string: serverURL.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            throw TimingAPIError.invalidServerURL
        }
        return base.appending(path: path)
    }

    private func validate(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else {
            throw TimingAPIError.requestFailed("No HTTP response from timing server.")
        }
        guard 200..<300 ~= http.statusCode else {
            if let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let detail = body["detail"] as? String {
                throw TimingAPIError.serverRejected(detail)
            }
            throw TimingAPIError.serverRejected("Timing server rejected the request (\(http.statusCode)).")
        }
    }
}
